contributor-docs/code-change-guide.md

```diff
@@ -0,0 +1,516 @@
+
+
+### Building containers for modified SDK code
+
+1. Command:
+  ```shell
+  ./gradlew :sdks:python:container:py39:docker \
+  -Pdocker-repository-root=<gcr.io/location> -Pdocker-tag=<tag>
+  ```
+2. Push containers
+3. Specify location with `--sdk_container_image`
+
+Example invocation command:
+
+```shell
+python -m pytest  -o log_cli=True -o log_level=Info \
+  apache_beam/ml/inference/pytorch_inference_it_test.py::PyTorchInference \
+  --test-pipeline-options='--runner=TestDataflowRunner --project=<project>
+                           --temp_location=gs://<bucket>/tmp
+                           --sdk_container_image=us.gcr.io/apache-beam-testing/beam-sdk/beam:dev
+                           --region=us-central1’
+```
+
+### Specifying additional test dependencies
+
+Option 1: Use the --requirements_file flag, e.g.
+
+```shell
+python -m pytest  -o log_cli=True -o log_level=Info \
+  apache_beam/ml/inference/pytorch_inference_it_test.py::PyTorchInference \
+  --test-pipeline-options='--runner=TestDataflowRunner --project=<project>
+                           --temp_location=gs://<bucket>/tmp
+                           --sdk_location=us.gcr.io/apache-beam-testing/beam-sdk/beam:dev
+                           --region=us-central1
+                           –requirements_file=requirements.txt’
+```
+
+Option 2: (for Dataflow runner) Use [custom containers](https://cloud.google.com/dataflow/docs/guides/using-custom-containers)
+
+Custom container can be based on official Beam SDK container image (See Appendix for url)
+
+## Run your pipeline with modified beam code
+
+1. Firstly, build Beam SDK tarball as instructed above (under `sdks/python`, run `python -m build –sdist` )
+
+2. Install the Beam SDK under your Python virtual environment with necessary extensions (e.g. `pip install /path/to/apache-beam.tar.gz[gcp]`)
+
+3. Initiate your Python script running the Pipeline with command like
+  ```shell
+  python my_pipeline.py --runner=DataflowRunner --sdk_location=/path/to/apache-beam.tar.gz --project=my_project --region=us-central1 --temp_location=gs://my-bucket/temp ...
+  ```
+
+Tips for Dataflow runner:
+
+* Python worker will install the provided Apache Beam SDK before processing work items, so usually no need to provide custom worker container unless your GCP VM does not have Internet access and transient dependencies are changed from the officially released container images. In this case, refer to "Building containers for modified SDK code" instruction above
+
+* Installing Beam Python SDK from source is known to be slow (3.5 min for `n1-standard-1` machine). Alternatively, if the host machine is amd64 architecture, one can build a wheel (instead of a tarball) with command with command (e.g. `./gradle :sdks:python:bdistPy311linux` for Python 3.11) and pass the built wheel to `--sdk_location`. The installation should then just take seconds.
```

Comment:
```suggestion
* Installing the Beam Python SDK from source can be slow (3.5 minutes for  a`n1-standard-1` machine). As an alternative, if the host machine uses amd64 architecture, you can build a wheel instead of a tarball by using a command similar to `./gradle :sdks:python:bdistPy311linux` (for Python 3.11). Pass the built wheel using the `--sdk_location` option. That installation completes in seconds.
```